### PR TITLE
Compass component

### PR DESCRIPTION
**Pull request by arcturus**
Adding the compass component.

Also added a debug scene that is loaded just in debug builds (accessible from the Settings scene). That view has a version of the compass pointing towards the shard. Also there you can see how the component is used (it will require to know what's the latitude and longitude of your location and the point you want to point to).

Considerations:
- In ios sensors works beautifully and there is no problems
- In android sensors may take a while until having a good reading and you will see how the arrow moves slowly until we get real data. We can 'calibrate' the compass by asking the user to do an '8' with the phone, it's unfortunately something that happens in android apps. Right now we don't detect the accuracy of the reading, but we could, in order to show a message like that this: http://uploads.tapatalk-cdn.com/20150824/8367bca44f8cca19edfa97d5092fce99.jpg

Created an app to try out several of this compasses working at the same time and works beautifully:
img_0001

Here is the repo if you want to play with it:
https://github.com/arcturus/ReactnativeCompass

@gmarty @wilsonpage r?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
   "dependencies": {
     "react": "~15.4.0-rc.4",
     "react-native": "0.40.0",
+    "react-native-heading": "git://github.com/arcturus/react-native-heading.git",
```
Comment (wilsonpage) [inline package.json:15]: What changes does your fork contain?

Comment (arcturus) [reply package.json:15]: I need it to be compatible with react 0.40 for iOS, did a PR to the original repo but after 1 week has not been merged.

Also the guy has a pending PR that I've checked to be good, so I added to the fork.
The rest is the same.

File: js/components/Compass/index.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+Compass.propTypes = {
+  azimut: React.PropTypes.number,
+  degreesFilter: React.PropTypes.number,
+  fromLat: React.PropTypes.number,
+  fromLon: React.PropTypes.number,
+  toLat: React.PropTypes.number,
+  toLon: React.PropTypes.number
+};
+
+module.exports = Compass;
```
Comment (wilsonpage) [inline js/components/Compass/index.js]: You're exporting the class twice. See above (line9).

File: js/scenes/Debug.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      , (error) => console.log(JSON.stringify(error)), 
+      {enableHighAccuracy: true, timeout: 20000, maximumAge: 1000}
+    ); 
+    this.watchID = navigator.geolocation.watchPosition((position) => { 
+      this.setState({
+        latitude: position.coords.latitude,
+        longitude: position.coords.longitude 
+      }); 
+    }); 
+}
```
Comment (wilsonpage) [inline js/scenes/Debug.js]: NIT: Indentation

File: js/components/Compass/index.js (hunk 2 above)
Comment (arcturus) [reply js/components/Compass/index.js]: Do'h! Good catch.

File: js/scenes/Settings.js [hunk 4]
```diff
@@ -31,7 +31,12 @@ export default class Settings extends Component {
     });
   }
 
+  onNavigate(id) {
```
Comment (wilsonpage) [inline js/scenes/Settings.js]: NIT: Naming: This isn't a response/event to navigation but an instruction to navigate.

`onNavigate` -> `navigate`

File: js/scenes/Settings.js [hunk 5]
```diff
@@ -45,13 +50,26 @@ export default class Settings extends Component {
...
+
+  renderDebug() {
+    if (__DEV__) {
+      return (<TouchableOpacity
+          onPress={this.onNavigate.bind(this, 'debug')}>
+          <Text style={styles.text}>Go to Debug</Text>
+        </TouchableOpacity>);
+    } else {
+      return '';
+    }
```
Comment (wilsonpage) [inline js/scenes/Settings.js:68]: NIT: Unnecessary `else`.

```js
renderDebug() {
  if (!__DEV__) return;
  return <TouchableOpacity onPress={this.onNavigate.bind(this, 'debug')}>
     <Text style={styles.text}>Go to Debug</Text>
  </TouchableOpacity>;
}
```

File: js/scenes/Debug.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      longitude: 0
+    };
+  }
+
+  onBackPress() {
+    this.navigator.pop();
+  }
+
+  // Watch for position for testing the compass
+  componentWillUnmount() {
```
Comment (wilsonpage) [inline js/scenes/Debug.js]: NIT: Can this be after `componentDidMount` so we see setup then teardown when reading.

File: js/scenes/Debug.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    this.navigator.pop();
+  }
+
+  // Watch for position for testing the compass
+  componentWillUnmount() { 
+    navigator.geolocation.clearWatch(this.watchID); 
+  }
+
+  componentDidMount() { 
+    navigator.geolocation.getCurrentPosition( (position) => {
```
Comment (wilsonpage) [inline js/scenes/Debug.js:31]: I think eventually the user's current position should be handled globally by Redux. This component could be dumber and just be passed a `deviceLocation` prop.

File: js/components/Compass/index.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+      outputRange: ['0 deg', '360 deg']
+    })}]}]}
+      source={require('./assets/arrow.png')}
+    />);  
+  }
+}
+
+Compass.propTypes = {
+  azimut: React.PropTypes.number,
+  degreesFilter: React.PropTypes.number,
```
Comment (wilsonpage) [inline js/components/Compass/index.js:110]: What do `azimut` and `degreesFilter` mean?

File: js/components/Compass/index.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    this.state.angleFilter = parseInt(filter);
+    this.stopHeadingUpdates();
+    this.startHeadingUpdates();
+  }
+
+  startHeadingUpdates() {
+    ReactNativeHeading.start(this.state.angleFilter)
+    .then(didStart => {
+        this.setState({
+            headingIsSupported: didStart,
```
Comment (wilsonpage) [inline js/components/Compass/index.js]: NIT: indentation

File: ios/MagnetJourney/Info.plist [hunk 10]
```diff
@@ -50,10 +33,26 @@
...
+		<string>armv7</string>
+	</array>
+	<key>UISupportedInterfaceOrientations</key>
+	<array>
+		<string>UIInterfaceOrientationPortrait</string>
+		<string>UIInterfaceOrientationLandscapeLeft</string>
+		<string>UIInterfaceOrientationLandscapeRight</string>
+	</array>
+	<key>UIViewControllerBasedStatusBarAppearance</key>
+	<false/>
```
Comment (gmarty) [inline ios/MagnetJourney/Info.plist]: Most of the stuff has just been reorganised?
It's hard to tell what changed.

File: js/components/Compass/index.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  setFilter(filter) {
+    this.state.angleFilter = parseInt(filter);
+    this.stopHeadingUpdates();
+    this.startHeadingUpdates();
+  }
+
+  startHeadingUpdates() {
+    ReactNativeHeading.start(this.state.angleFilter)
```
Comment (wilsonpage) [inline js/components/Compass/index.js:32]: If we have lots of these components, does it matter if `start` is called several times? Again I think eventually state like this should be passed in and managed by Redux.

File: js/components/Compass/index.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
+import React, { Component } from 'react';
+import {
+  Image,
+  Animated,
+  DeviceEventEmitter
+} from 'react-native';
+import ReactNativeHeading from 'react-native-heading';
+
+export default class Compass extends Component {
+
```
Comment (gmarty) [inline js/components/Compass/index.js]: nit: no line break. See other files for coding style.
You can also run `yarn run fix` to fix potential basic issues with styling.

File: js/components/Compass/index.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  setFilter(filter) {
+    this.state.angleFilter = parseInt(filter);
+    this.stopHeadingUpdates();
+    this.startHeadingUpdates();
+  }
+
+  startHeadingUpdates() {
+    ReactNativeHeading.start(this.state.angleFilter)
+    .then(didStart => {
```
Comment (gmarty) [inline js/components/Compass/index.js]: nit: this block should be indented.

File: js/components/Compass/index.js [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    const radiansBearing = Math.atan2(y, x);
+
+    return parseInt(this.toDegrees(radiansBearing));
+  }
+
+  render() {
+    return (<Animated.Image style={[{transform: [{rotate: this.state.rotation.interpolate({
+      inputRange:[0, 360],
+      outputRange: ['0 deg', '360 deg']
+    })}]}]}
```
Comment (wilsonpage) [inline js/components/Compass/index.js]: NIT: Kinda hard to parse this line.

File: js/components/Compass/index.js (hunk 13 above)
Comment (gmarty) [reply js/components/Compass/index.js]: what about:
```js
.then(headingIsSupported => {
  this.setState({ headingIsSupported })
});
```

File: js/scenes/Settings.js (hunk 5 above)
Comment (arcturus) [reply js/scenes/Settings.js:68]: That's horrendous! Difficult to see there are two paths in that function :P

File: js/scenes/Debug.js (hunk 7 above)
Comment (arcturus) [reply js/scenes/Debug.js:31]: Yes, I think we will have an unique source of truth for the location, but until we have that, (we marked as one of the components), I was using that cheap way of having location for testing the component.

File: js/components/Compass/index.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    const toLat = this.toRadians(this.props.toLat);
+    const toLon = this.toRadians(this.props.toLon);
+
+    const dLon = toLon - fLon;
+
+    const y = Math.sin(dLon) * Math.cos(toLat);
+    const x = Math.cos(fLat) * Math.sin(toLat) - Math.sin(fLat) * Math.cos(toLat) * Math.cos(dLon);
+    const radiansBearing = Math.atan2(y, x);
+
+    return parseInt(this.toDegrees(radiansBearing));
```
Comment (gmarty) [inline js/components/Compass/index.js:89]: is `parseInt` neeeded here?

File: js/scenes/Settings.js (hunk 5 above)
Comment (wilsonpage) [reply js/scenes/Settings.js:68]: Well it's essentially a `guard`. If we're not in DEV mode it's a noop. Francisco style:

```js
renderDebug() {
  if (!__DEV__) {
    return;
  }

  return <TouchableOpacity onPress={this.onNavigate.bind(this, 'debug')}>
     <Text style={styles.text}>Go to Debug</Text>
  </TouchableOpacity>;
}
```

File: js/components/Compass/index.js (hunk 11 above)
Comment (arcturus) [reply js/components/Compass/index.js:32]: It doesn't matter, same than gps is a sensor that is being already used, and we will listen to the updated values.

File: js/scenes/Debug.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+Debug.propTypes = {
+  navigator: PropTypes.object,
+};
+
+const styles = StyleSheet.create({
+  text: {
+    ...defaultTextStyle,
+  },
+});
```
Comment (gmarty) [inline js/scenes/Debug.js]: Add a line break at the end of files.

File: js/components/Compass/index.js (hunk 8 above)
Comment (arcturus) [reply js/components/Compass/index.js:110]: azimut ... damn! I used the Spanish and French version, in English is azimuth, it's pretty much the naming you use to refer to what's your reference, north, when computing locations, angles and so on. But it's true I need to rename it to proper English.

File: js/scenes/Settings.js (hunk 4 above)
Comment (gmarty) [reply js/scenes/Settings.js]: I'd say `onNavigationChange` though to make the relation with a user interaction more obvious.

File: js/components/Compass/index.js [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    this.startHeadingUpdates();
+  }
+
+  componentWillUnmount() {
+    this.stopHeadingUpdates();
+  }
+
+  startAnimation(z) {
+    const offset = this.calculateOffset();
+    if (offset != 0) {
```
Comment (gmarty) [inline js/components/Compass/index.js]: Surely it won't hurt if you subtract 0.

File: js/components/Compass/index.js (hunk 8 above)
Comment (wilsonpage) [reply js/components/Compass/index.js:110]: Oh, well I've never heard of either :p

**Review by gmarty — CHANGES_REQUESTED** [review]
Looks good to me.
The trigonometric computations works for us but might not at the scale of earth.

A few comments left but ship it!!

File: js/scenes/Debug.js (hunk 6 above)
Comment (arcturus) [reply js/scenes/Debug.js]: Yup, good catch.

**Comment by wilsonpage** [comment]
I need to test on device when I get into the office but code looks good to me. Although I don't know anything about trigonometry, angles and all that jazz. Seems @gmarty knows more about that :)

File: js/components/Compass/index.js (hunk 8 above)
Comment (arcturus) [reply js/components/Compass/index.js:110]: degreesFilter is the number of degrees your heading has to change in order to notify of a change.

The hardware is going to be constantly giving values, different values, even if we are quiet, so we need to have a threshold to not to make the arrow shake too much.

File: js/components/Compass/index.js (hunk 14 above)
Comment (arcturus) [reply js/components/Compass/index.js]: Will try my best, but believe me the syntax for interpolation in animations is complicated.

File: js/scenes/Settings.js (hunk 5 above)
Comment (arcturus) [reply js/scenes/Settings.js:68]: Ok, I can do that :) ... we should use more Francisco's style.

File: ios/MagnetJourney/Info.plist (hunk 10 above)
Comment (arcturus) [reply ios/MagnetJourney/Info.plist]: Those are automatic changes done by xcode :S

File: js/components/Compass/index.js (hunk 17 above)
Comment (arcturus) [reply js/components/Compass/index.js]: Yes it wont :), also the checking of the latitude and longitude is happening in calculateOffset so is safe to skip that check.

File: js/components/Compass/index.js (hunk 15 above)
Comment (arcturus) [reply js/components/Compass/index.js:89]: Yes, we need to return integers, so the animation interpolation can go from 0 deg to 360 deg and the animation comes for free in react native.

File: js/components/Compass/index.js [hunk 18]
```diff
@@ -1,8 +1,10 @@
+/* global require */
```
Comment (wilsonpage) [inline js/components/Compass/index.js:1]: Why is this needed? If we support `require()` syntax can't we define that globally in the app's `.eslintrc`?

**Comment by wilsonpage** [comment]
Having trouble building on Android:

screen shot 2017-01-18 at 17 21 53

**Comment by wilsonpage** [comment]
I cast to `int` and seems to be working :-/

```java
newAzimuth = (int) ...
```

Rotation looks nice and smooth. Really cool to see this working!

**Comment by wilsonpage** [comment]
@arcturus so not sure what do do about the compiler error in `react-native-heading`. Strange that you don't have it.

**Comment by wilsonpage** [comment]
I tried to test on IOS but I think I'm missing some kind of certificate to build on a real device. Help needed :)

**Comment by wilsonpage** [comment]
Debug menu is missing a back button.

**Comment by arcturus** [comment]
Good catch with the android version, had it modified locally but didn't push it.

**Comment by arcturus** [comment]
Added back button to the debug build.

**Comment by arcturus** [comment]
I did rebase, but seems that now the react-native-heading is not compiling in the new workspace :(

**Comment by arcturus** [comment]
Finally found the problem, it took me a whole day.

With the new cocoapods, in order to compiile the library libreactheading we need to change the search for the headers library to where the headers are now (after using pods):
$(SRCROOT)/../../ios/Pods/Headers/Private/React/**

😭 😭 😭 😭 😭 😭 😭

**Comment by arcturus** [comment]
Had to change to npm in CI since yarn is hanging while downloading packages:

https://github.com/yarnpkg/yarn/issues/764
